packages/babel-helper-builder-react-jsx-experimental/src/index.js

```diff
@@ -0,0 +1,844 @@
+          state.set("pragmaSet", pragmaSet);
+          state.set("pragmaFragSet", pragmaFragSet);
+        } else if (runtime === "automatic") {
+          const importName = addAutoImports(path, {
+            ...state.opts,
+            source,
+          });
+          state.set(
+            "jsxIdentifier",
+            createIdentifierParser(
+              createIdentifierName(
+                path,
+                options.development ? "jsxDEV" : "jsx",
+                importName,
+              ),
+            ),
+          );
+
+          state.set(
+            "jsxStaticIdentifier",
+            createIdentifierParser(
+              createIdentifierName(
+                path,
+                options.development ? "jsxDEV" : "jsxs",
+                importName,
+              ),
+            ),
+          );
+
+          state.set(
+            "createElementIdentifier",
+            createIdentifierParser(
+              createIdentifierName(path, "createElement", importName),
+            ),
+          );
+
+          state.set(
+            "jsxFragIdentifier",
+            createIdentifierParser(
+              createIdentifierName(path, "Fragment", importName),
+            ),
+          );
+        } else {
+          throw path.buildCodeFrameError(
+            `Runtime must be either "classic" or "automatic".`,
+          );
+        }
+      }
+    },
+
+    exit(path, state) {
+      if (
+        state.get("runtime") === "classic" &&
+        state.get("pragmaSet") &&
+        state.get("usedFragment") &&
+        !state.get("pragmaFragSet")
+      ) {
+        throw new Error(
+          "transform-react-jsx: pragma has been set but " +
+            "pragmaFrag has not been set",
+        );
+      }
+    },
+  };
+
+  return visitor;
+
+  // We want to use React.createElement, even in the case of
+  // jsx, for <div {...props} key={key} /> to distinguish it
+  // from <div key={key} {...props} />. This is an intermediary
+  // step while we deprecate key spread from props. Afterwards,
+  // we will stop using createElement in the transform.
+  function shouldUseCreateElement(path) {
+    const openingPath = path.get("openingElement");
+    const attributes = openingPath.node.attributes;
+
+    let seenPropsSpread = false;
+    for (let i = 0; i < attributes.length; i++) {
+      const attr = attributes[i];
+      if (
+        seenPropsSpread &&
+        t.isJSXAttribute(attr) &&
+        attr.name.name === "key"
+      ) {
+        return true;
+      } else if (t.isJSXSpreadAttribute(attr)) {
+        seenPropsSpread = true;
+      }
+    }
+    return false;
+  }
+
+  function createIdentifierName(path, name, importName) {
+    if (isModule(path) && importName) {
+      const identifierName = `${importName[name]}`;
+      return identifierName;
+    } else {
+      return `${importName}.${name}`;
+    }
+  }
+
+  function getImportNames(parentPath) {
+    const imports = {};
```

Comment:
Since this is only used to set properties as `true`, and then in `addAutoImports` we don't even check what the value of the properties is but only their presence, it seems the perfect use case for a `new Set()`.